system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -372,6 +393,396 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>
+            xpath+=20;
+            node = xmlDocGetRootElement(m_xmlDoc);
+        }
+
+        StringArray xpathnodes;
+        xpathnodes.appendList(xpath, "/");
+        ForEachItemIn(i, xpathnodes)
+        {
+            const char *finger = xpathnodes.item(i);
+            if (isEmptyString(finger)) //shouldn't happen, scripting error rather than not found, required==false has no effect
+                throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: '//' not allowed '%s'", xpath);
+            if (*finger=='@')
+                throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: attribute cannot be selected '%s'", xpath);
+
+            xmlXPathObjectPtr obj = xmlXPathNodeEval(node, (const xmlChar*) finger, m_xpathContext);
+            xmlXPathSetContextNode(current, m_xpathContext); //restore
+
+            if (!obj)
+                throw MakeStringException(XPATHERR_InvalidInput,"XpathContext:ensureLocation: invalid xpath '%s' at node '%s'", xpath, finger);
+
+            if (obj->type!=xmlXPathObjectType::XPATH_NODESET || !obj->nodesetval || obj->nodesetval->nodeNr==0)
+            {
+                if (*finger=='.' || strpbrk(finger, "[()]*") || strstr(finger, "::")) //complex nodes must already exist
+                {
+                    if (required)
+                        throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: xpath node not found '%s' in xpath '%s'", finger, xpath);
+                    return false;
+                }
+                node = xmlNewChild(node, nullptr, (const xmlChar *) finger, nullptr);
+                if (!node)
+                {
+                    if (required)
+                        throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: error creating node '%s' in xpath '%s'", finger, xpath);
+                    return false;
+                }
+            }
+            else
+            {
+                xmlNodePtr *nodes = obj->nodesetval->nodeTab;
+                if (obj->nodesetval->nodeNr>1)
+                {
+                    if (required)
+                        throw MakeStringException(XPATHERR_InvalidState,"XpathContext:ensureLocation: ambiguous xpath '%s' at node '%s'", xpath, finger);
+                    return false;
+                }
+                node = *nodes;
+            }
+            xmlXPathFreeObject(obj);
+        }
+        setLocation(node);
+        return true;
+    }
+
+    virtual void setLocationNamespace(const char *prefix, const char * uri, bool current) override
+    {
+        xmlNodePtr node = m_xpathContext->node;
+        if (!node)
+            return;
+        xmlNsPtr ns = nullptr;
+        if (isEmptyString(uri))
+            ns = xmlSearchNs(node->doc, node, (const xmlChar *) prefix);
+        else
+            ns = xmlNewNs(node, (const xmlChar *) uri, (const xmlChar *) prefix);
```

Comment:
What if the prefix/uri is a redefinition? Will xmlNewNs throw an error? Or what if the prefix is reused with a different uri? Asking mostly because some of the libxml errors aren't very friendly in that they don't  show up in the esp.log file.

Reply:
Right, I don't think you'd get an error redefining a prefix.  And routines relying on it would find the nearest location where the prefix was defined.  libxml errors and implementation can be a bit quirky though.